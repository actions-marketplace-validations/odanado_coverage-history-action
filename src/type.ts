export interface Position {
  line: number;
  column: number;
}

export interface Location {
  start: Position;
  end: Position;
}

export interface StatementMap {
  [index: string]: Location;
}

export interface FnMap {
  [index: string]: {
    name: string;
    decl: Location;
    loc: Location;
  };
}

export interface BranchMap {
  [inedx: string]: {
    loc: Location;
    type: string;
    locations: Location[];
  };
}

export interface Coverage {
  [path: string]: {
    path: string;
    statementMap: StatementMap;
    fnMap: FnMap;
    branchMap: BranchMap;
    s: {
      [index: string]: number;
    };
    f: {
      [index: string]: number;
    };
    b: {
      [index: string]: [number, number];
    };
  };
}
