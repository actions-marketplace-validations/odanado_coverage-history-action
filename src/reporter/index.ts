import { CoverageResult } from "../type";

export interface Reporter {
  report(
    targetCoverage: CoverageResult,
    currentCoverage: CoverageResult
  ): Promise<string>;
}
