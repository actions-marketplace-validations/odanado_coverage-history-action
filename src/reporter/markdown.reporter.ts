import { Reporter } from "./index";
import { CoverageSet } from "../type";

export function tableMaker(
  strings: TemplateStringsArray,
  ...placeholders: string[]
): (values: { [key: string]: string }) => string {
  return (values: { [key: string]: string }): string => {
    return (
      strings[0] +
      placeholders
        .map((placeholder, i) => {
          return `${values[placeholder]}${strings[i + 1]}`;
        })
        .join("")
        .trim()
    );
  };
}

export class MarkdownReporter implements Reporter {
  private format(coverage: number): string {
    return (coverage * 100).toFixed(2);
  }
  report(
    targetCoverage: CoverageSet,
    currentCoverage: CoverageSet
  ): Promise<string> {
    const makeTable = tableMaker`
|  | target | current |
| - | -: | -: |
|  statement  | ${"targetStatement"} | ${"currentStatement"} |
|  branch  | ${"targetBranch"} | ${"currentBranch"} |
|  function  | ${"targetFunction"} | ${"currentFunction"} |
`;

    const table = makeTable({
      targetStatement: `${this.format(targetCoverage.statement.total)} %`,
      currentStatement: `${this.format(currentCoverage.statement.total)} %`,
      targetBranch: `${this.format(targetCoverage.branch.total)} %`,
      currentBranch: `${this.format(currentCoverage.branch.total)} %`,
      targetFunction: `${this.format(targetCoverage.function.total)} %`,
      currentFunction: `${this.format(currentCoverage.function.total)} %`,
    });

    return Promise.resolve(table);
  }
}
