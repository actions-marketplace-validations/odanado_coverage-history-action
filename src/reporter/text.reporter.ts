import { average } from "../utils/average";

import { Reporter } from "./index";
import { Coverage } from "../type";

export class TextReporter implements Reporter {
  // TODO: 別ファイル？
  calcStatementCoverage(coverage: Coverage): number {
    const paths = Object.keys(coverage);
    const sumList = paths.map((path) => {
      return Object.values(coverage[path].s).reduce(
        (prev, cur) => {
          if (cur > 0) {
            return [prev[0] + 1, prev[1] + 1];
          }
          return [prev[0], prev[1] + 1];
        },
        [0, 0]
      );
    });

    const summary = Object.fromEntries(
      paths.map((path, i) => {
        return [path, sumList[i][0] / sumList[i][1]];
      })
    );

    const total = sumList.reduce(
      (prev, cur) => {
        return [prev[0] + cur[0], prev[1] + cur[1]];
      },
      [0, 0]
    );

    return total[0] / total[1];
  }
  report(targetCoverage: Coverage, currentCoverage: Coverage): Promise<string> {
    const targetStatementCoverage = this.calcStatementCoverage(targetCoverage);
    const currentStatementCoverage = this.calcStatementCoverage(
      currentCoverage
    );
    return Promise.resolve(
      `target: ${targetStatementCoverage}\tcurrent: ${currentStatementCoverage}`
    );
  }
}
