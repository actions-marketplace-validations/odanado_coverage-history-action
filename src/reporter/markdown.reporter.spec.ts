import { MarkdownReporter } from "./markdown.reporter";
import { CoverageSet } from "../type";

describe("MarkdownReporter", () => {
  const target: CoverageSet = {
    statement: {
      total: 0.1,
      paths: [],
    },
    branch: {
      total: 0.2,
      paths: [],
    },
    function: {
      total: 0.3,
      paths: [],
    },
  };
  const current: CoverageSet = {
    statement: {
      total: 0.4,
      paths: [],
    },
    branch: {
      total: 0.5,
      paths: [],
    },
    function: {
      total: 0.6,
      paths: [],
    },
  };
  it("report correctly", async () => {
    const reporter = new MarkdownReporter();
    const report = await reporter.report(
      { branch: "master", coverage: target },
      { branch: "feature", coverage: current }
    );
    expect(report).toMatchSnapshot();
  });
});
