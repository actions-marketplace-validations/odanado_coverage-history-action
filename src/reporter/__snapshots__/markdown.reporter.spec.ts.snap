// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MarkdownReporter report correctly 1`] = `
"
|  | target | current |
| - | -: | -: |
|  statement  | 10.00 % | 40.00 % |
|  branch  | 20.00 % | 50.00 % |
|  function  | 30.00 % | 60.00 % |"
`;
